feat: add `console/log-each-map`

headlessNode opened the pull request:
## Description

> What is the purpose of this pull request?

This pull request:

-   add `@stdlib/console/logEachMap` with relevant documentation, benchmarks, tests and examples.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves https://github.com/stdlib-js/metr-issue-tracker/issues/13 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

headlessNode committed:
feat: add console/logEachMap

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/console/log-each-map/lib/main.js:
```diff
@@ -0,0 +1,162 @@
+	}
+
+	for ( i = 0; i < len; i++ ) {
+		for ( j = 0; j < n; j++ ) {
+			values[ j+1 ] = getters[ j ]( args[ j ], offsets[ j ] );
+			offsets[ j ] += strides[ j ];
+		}
+		cbArgs = slice( values, 1, n+1 );
+		cbArgs.push( i );
+		for ( j = 0; j < args.length; j++ ) {
+			cbArgs.push( args[j] );
+		}
```
This logic needs to be revisited. You should be able to preallocate `cbArgs`, rather than slice and copy. In the loop at L145, you can assign to an element in `cbArgs`. As for the arrays which are passed to the `clbk`, prefer the signature `clbk( v1, v2, ..., index, arrays )`, rather than `clbk( v1, v2, ..., index, arr1, arr2, ... )`.

kgryte replied:
In your preallocated `cbArgs`, the `arrays` argument never needs to be updated as you are doing in L151, L152, as the list of arrays never changes.

headlessNode replied:
Makes sense.

headlessNode committed:
refactor: apply suggestions from code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

kgryte commented on lib/node_modules/@stdlib/console/log-each-map/lib/main.js:
```diff
@@ -0,0 +1,160 @@
+		offsets.push( 0 );
+	}
+
+	cbArgs = [];
+	cbArgs[ n + 1 ] = args;
```
Don't do this. This automatically puts arrays into dictionary mode. You need to explicitly allocate to ensure a contiguous generic array. Feel free to use https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/array/base/nulls.

kgryte commented on lib/node_modules/@stdlib/console/log-each-map/README.md:
```diff
@@ -0,0 +1,183 @@
+
+The function accepts the following arguments:
+
+-   **str**: Format string.
```
```suggestion
-   **str**: format string.
```

Lowercase descriptions here and below.

kgryte commented on lib/node_modules/@stdlib/console/log-each-map/docs/repl.txt:
```diff
@@ -0,0 +1,30 @@
+{{alias}}( str[, ...args], clbk[, thisArg] )
```
Missing empty line.

kgryte commented on lib/node_modules/@stdlib/console/log-each-map/docs/types/index.d.ts:
```diff
@@ -0,0 +1,131 @@
+/**
+* Inserts array element values into a format string, applies a callback, and prints the result.
+*
+* @param str - Format string
```
Lowercase descriptions here and everywhere.

kgryte commented on lib/node_modules/@stdlib/console/log-each-map/docs/types/index.d.ts:
```diff
@@ -0,0 +1,131 @@
+* @param clbk - Callback function
+* @param thisArg - Callback execution context
+*/
+declare function logEachMap( str: string, clbk: NullaryCallback, thisArg?: any ): void;
```
These declarations can be significantly improved to better preserve type information.

kgryte commented on lib/node_modules/@stdlib/console/log-each-map/docs/types/index.d.ts:
```diff
@@ -0,0 +1,131 @@
+type NullaryCallback = () => any;
+
+/**
+* Callback function that accepts one argument.
```
You need to document parameters and return values.

kgryte commented on lib/node_modules/@stdlib/console/log-each-map/docs/types/index.d.ts:
```diff
@@ -0,0 +1,131 @@
+/**
+* Callback function that accepts no arguments.
+*/
+type NullaryCallback = () => any;
```
You should be carrying forward the `this` argument.

headlessNode committed:
refactor: apply suggestions from code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

kgryte commented on lib/node_modules/@stdlib/console/log-each-map/docs/repl.txt:
```diff
@@ -0,0 +1,31 @@
+    --------
+    > var x = [ 1, 2, 3 ];
+    > var y = [ 4, 5, 6 ];
+    > function sum( x, y ) { return x + y; };
+    > {{alias}}( '%d + %d = %d', x, y, sum );
```
```suggestion
    > function f( x, y ) { return x + y; };
    > {{alias}}( '%d + %d = %d', x, y, f );
```

Try to avoid using function names which will likely have existing global aliases.

headlessNode replied:
This applies to the `repl` only or elsewhere too?

kgryte replied:
Only the REPL, as there we have a global namespace.

headlessNode committed:
docs: apply suggestion from code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

kgryte commented on lib/node_modules/@stdlib/console/log-each-map/lib/main.js:
```diff
@@ -0,0 +1,161 @@
+	}
+
+	cbArgs = nulls( n + 2 );
+	cbArgs[ n + 1 ] = args;
+	for ( i = 0; i < len; i++ ) {
+		for ( j = 0; j < n; j++ ) {
+			cbArgs[ j ] = getters[ j ]( args[ j ], offsets[ j ] );
+			values[ j + 1 ] = cbArgs[ j ];
+			offsets[ j ] += strides[ j ];
+		}
+		cbArgs[ n ] = i;
+		values[ n + 1 ] = clbk.apply( thisArg, cbArgs );
```
```suggestion
	cbArgs[ n+1 ] = args;
	for ( i = 0; i < len; i++ ) {
		for ( j = 0; j < n; j++ ) {
			cbArgs[ j ] = getters[ j ]( args[ j ], offsets[ j ] );
			values[ j+1 ] = cbArgs[ j ];
			offsets[ j ] += strides[ j ];
		}
		cbArgs[ n ] = i;
		values[ n+1 ] = clbk.apply( thisArg, cbArgs );
```

kgryte committed:
style: reduce whitespace noise

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
chore: clean-up

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

kgryte committed:
test: update tests

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte committed:
docs: update descriptions

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte committed:
docs: update description

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte committed:
docs: update description

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte reviewed (approved):
LGTM after refactoring and clean-up.

kgryte commented:
@headlessNode I suggest taking a look at how I refactored the TypeScript declarations. The original declarations were not good about preserving type specificity.
